build/helper/codegen_helper.py

```diff
@@ -131,108 +137,138 @@ def _get_buffer_parameter_for_size_parameter(parameter, parameters):
-    module_name = 'visatype'
-    custom_type = find_custom_type(parameter, config)
-    if custom_type is not None:
-        module_name = custom_type['file_name']
+    assert parameter['is_buffer'] is False
+    assert parameter['numpy'] is False
+    corresponding_buffer_parameter = _get_buffer_parameter_for_size_parameter(parameter, parameters)
 
-    # And now: A large block of conditional logic for getting to each of these cases. The following will not win any beauty pageants.
-    # Suggestions on how to improve readability are welcome.
-    # Note that we append "# case x". It's ugly in the generated code but it's sooo useful for debugging code generation problems.
-    if parameter['size']['mechanism'] == 'python-code':
-        size = parameter['size']['value']
-        # Now we need to replicate some of the same conditions from below
-        # We only support non-buffer/array in parameters or buffer/array out parameters
-        if parameter['direction'] == 'in':
-            assert parameter['is_buffer'] is False
-            definition = '{0}.{1}({2})  # case 0.0'.format(module_name, parameter['ctypes_type'], size)
-        else:
-            assert parameter['is_buffer'] is True
-            assert parameter['direction'] == 'out'
-            definition = '({0}.{1} * {2})()  # case 0.2'.format(module_name, parameter['ctypes_type'], size)
-    elif parameter['direction'] == 'in':
+    if parameter['direction'] == 'in':
         if parameter['is_session_handle'] is True:
             definition = '{0}.{1}(self._{2})  # case 1'.format(module_name, parameter['ctypes_type'], parameter['python_name'])
-        elif parameter['is_repeated_capability'] is True:
+        elif parameter['size']['mechanism'] == 'python-code':
+            definition = '{0}.{1}({2})  # case 0.0'.format(module_name, parameter['ctypes_type'], parameter['size']['value'])
+        elif parameter['enum'] is not None:
+            definition = '{0}.{1}({2}.value)  # case 10'.format(module_name, parameter['ctypes_type'], parameter['python_name'])
+        elif corresponding_buffer_parameter is None:
+            definition = '{0}.{1}({2})  # case 9'.format(module_name, parameter['ctypes_type'], parameter['python_name'])
+        elif corresponding_buffer_parameter['direction'] == 'in':
+            # Parameter denotes the size of another (the "corresponding") parameter.
+            definition = '{0}.{1}(0 if {2} is None else len({2}))  # case 6'.format(module_name, parameter['ctypes_type'], corresponding_buffer_parameter['python_name'])
+        else:
+            assert corresponding_buffer_parameter['direction'] == 'out'
+            if corresponding_buffer_parameter['size']['mechanism'] == 'ivi-dance':
+                assert ivi_dance_step == 1 or ivi_dance_step == 2, "ivi-dance has two steps, check metadata for parameter {0}".format(parameter['name'])
```

Comment:
Would it make sense to create an enum for ivi_dance_step to enforce its oneness or twoness?

Reply:
Done, helps with code readability.